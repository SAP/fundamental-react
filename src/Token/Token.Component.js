import React from 'react';
import { Token } from './Token';
import { Description, DocsText, DocsTile, Header, Import, Separator } from '../_playground';

export const TokenComponent = () => {
    const tokenCode = `<Token>Bibendum</Token>
<Token>Lorem</Token>
<Token>Dolor</Token>
<Token>Filter</Token>`;

    const closeAction = (name) => alert(`close ${name}! You can work with the event itself using the callback.`);
    const tagsNames = ['Bibendum', 'Lorem', 'Dolor', 'Filter'];

    return (
        <div>

            <Header>Token</Header>
            <Description>Tokens are used to represent contextualizing information. They can be useful to show applied filters, selected values for form field or object metadata.
            </Description>
            <Import module='Token' path='/fundamental-react/src/' />

            <Separator />

            <DocsTile centered>
                <div className='fd-doc__margin--token'>
                    {
                        tagsNames.map((name) => {
                            return <Token clickHandler={(e) => closeAction(name, e)}>{name}</Token>;
                        })
                    }
                </div>
            </DocsTile>
            <DocsText>{tokenCode}</DocsText>
        </div>
    );
};
