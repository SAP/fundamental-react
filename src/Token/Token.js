import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/token.css';

const classnames = classnamesBind.bind(styles);

/** A **Token** is used to represent contextual information. It can be useful to show
applied filters, selected values for a form field or object metadata. */
const Token = React.forwardRef(({
    buttonLabel,
    children,
    className,
    compact,
    cssNamespace,
    onClick,
    readOnly,
    ...props
}, ref) => {

    const tokenClasses = classnames(
        `${cssNamespace}-token`,
        {
            [`${cssNamespace}-token--readonly`]: readOnly,
            [`${cssNamespace}-token--compact`]: compact
        },
        className
    );

    return (
        <span
            {...props}
            className={tokenClasses}
            ref={ref}
            role='button'
            tabIndex='0'>
            <span className={classnames(`${cssNamespace}-token__text`)}>{children}</span>
            <button
                aria-label={buttonLabel}
                className={classnames(`${cssNamespace}-token__close`)}
                onClick={onClick}
                tabIndex='-1' />
        </span>
    );
});

Token.displayName = 'Token';

Token.propTypes = {
    /** A localized string to be used as aria-label for the token's button */
    buttonLabel: PropTypes.string.isRequired,
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to enable compact mode */
    compact: PropTypes.bool,
    /** Set to **true** to mark component as readonly */
    readOnly: PropTypes.bool,
    /**
     * Callback function; triggered when the Token's close `<button>` is clicked.
     *
     * @param {SyntheticEvent} event - React's original SyntheticEvent. See https://reactjs.org/docs/events.html.
     * @returns {void}
    */
    onClick: PropTypes.func
};

Token.defaultProps = {
    onClick: () => {}
};

export default withStyles(Token);
