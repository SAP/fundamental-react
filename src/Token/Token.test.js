import { mount } from 'enzyme';
import React from 'react';
import Token from './Token';

describe('<Token />', () => {
    describe('Prop spreading', () => {
        test('should allow props to be spread to the Token component', () => {
            const element = mount(<Token buttonLabel='placeholder label' data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });

    test('forwards the ref', () => {
        let ref;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                ref = React.createRef();
            }
            render = () => <Token buttonLabel='placeholder label' ref={ref} />;
        }
        mount(<Test />);
        expect(ref.current.tagName).toEqual('SPAN');
        expect(ref.current.className).toEqual('fd-token');
    });
});
