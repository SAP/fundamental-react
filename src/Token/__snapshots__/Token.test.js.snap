// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Token /> create token component 1`] = `
<span
  className="fd-token"
  role="button"
>
  Bibendum
</span>
`;

exports[`<Token /> create token component 2`] = `
<span
  className="fd-token blue"
  role="button"
>
  Bibendum
</span>
`;
